File: postlio_frontend/src/lib/api/ai.ts
// src/lib/api/ai.ts

import { apiClient } from './client';

// ============================================================
// ENUMS
// ============================================================

export type TextProvider = 'gemini' | 'groq';
export type ImageProvider = 'pollinations' | 'huggingface';
export type VideoProvider = 'pollinations';

export type Platform = 'facebook' | 'instagram' | 'linkedin';

export type Category =
    | 'fitness'
    | 'health'
    | 'beauty'
    | 'cooking'
    | 'business'
    | 'technology'
    | 'travel'
    | 'lifestyle'
    | 'education'
    | 'entertainment';

export type Tone =
    | 'professional'
    | 'casual'
    | 'humorous'
    | 'inspirational'
    | 'educational'
    | 'friendly';

export type ImageStyle =
    | 'realistic'
    | 'artistic'
    | 'cartoon'
    | 'minimalist'
    | 'vibrant'
    | 'professional';

// ============================================================
// PROVIDER INFO
// ============================================================

export interface VideoModelInfo {
    id: string;
    name: string;
}

export interface ProviderInfo {
    name: string;
    display_name?: string;
    available: boolean;
    is_free?: boolean;
    models: string[];
    models_detailed?: VideoModelInfo[];
    is_default: boolean;
    description?: string;
}

export interface ProvidersResponse {
    text_providers: ProviderInfo[];
    image_providers: ProviderInfo[];
    video_providers?: ProviderInfo[];
}

// ============================================================
// TEXT GENERATION
// ============================================================

export interface TextGenerationRequest {
    topic: string;
    platform: Platform;
    provider?: TextProvider;
    model?: string;
    category?: Category;
    tone?: Tone;
    language?: string;
    include_hashtags?: boolean;
    include_emoji?: boolean;
    max_length?: number;
}

export interface GeneratedTextContent {
    content: string;
    platform: string;
    provider: string;
    model?: string;
    hashtags?: string[];
}

export interface TextGenerationResponse {
    success: boolean;
    data: GeneratedTextContent;
    tokens_used?: number;
}

// ============================================================
// VARIATIONS
// ============================================================

export interface VariationsRequest {
    content: string;
    platform: Platform;
    provider?: TextProvider;
    model?: string;
    variations_count?: number;
}

export interface VariationsResponse {
    success: boolean;
    variations: string[];
    provider: string;
}

// ============================================================
// IMPROVE TEXT
// ============================================================

export interface ImproveRequest {
    content: string;
    platform: Platform;
    provider?: TextProvider;
    model?: string;
    instructions?: string;
}

export interface ImproveResponse {
    success: boolean;
    content: string;
    provider: string;
}

// ============================================================
// CHAT
// ============================================================

export interface ChatMessage {
    role: 'user' | 'assistant';
    content: string;
}

export interface ChatRequest {
    messages: ChatMessage[];
    provider?: TextProvider;
    model?: string;
    category?: Category;
    platform?: Platform;
}

export interface ChatResponse {
    success: boolean;
    message: string;
    provider: string;
}

// ============================================================
// IMAGE GENERATION
// ============================================================

export interface ImageGenerationRequest {
    prompt: string;
    provider?: ImageProvider;
    model?: string;
    style?: ImageStyle;
    width?: number;
    height?: number;
}

export interface GeneratedImageContent {
    image_url?: string;
    image_data?: string;
    prompt: string;
    prompt_translated?: string;
    prompt_enhanced?: string;
    provider: string;
    model?: string;
    width?: number;
    height?: number;
}

export interface ImageGenerationResponse {
    success: boolean;
    data?: GeneratedImageContent;
    error?: string;
}

// ============================================================
// VIDEO GENERATION
// ============================================================

export interface VideoGenerationRequest {
    prompt: string;
    provider?: VideoProvider;
    model?: string;
    width?: number;
    height?: number;
    duration?: number;
    reference_image?: string;
}

export interface GeneratedVideoContent {
    video_data?: string;
    mime_type?: string;
    prompt: string;
    prompt_translated?: string;
    provider: string;
    model?: string;
    model_display_name?: string;
    width?: number;
    height?: number;
    duration?: number;
    size_bytes?: number;
    has_reference_image: boolean;
}

export interface VideoGenerationResponse {
    success: boolean;
    data?: GeneratedVideoContent;
    error?: string;
}

// ============================================================
// API FUNCTIONS
// ============================================================

export async function getProviders(): Promise<ProvidersResponse> {
    return apiClient.get<ProvidersResponse>('/ai/providers');
}

export async function generateText(
    request: TextGenerationRequest
): Promise<TextGenerationResponse> {
    return apiClient.post<TextGenerationResponse>('/ai/generate/text', {
        topic: request.topic,
        platform: request.platform,
        provider: request.provider,
        model: request.model,
        category: request.category,
        tone: request.tone || 'professional',
        language: request.language || 'pl',
        include_hashtags: request.include_hashtags ?? true,
        include_emoji: request.include_emoji ?? true,
        max_length: request.max_length,
    });
}

export async function generateImage(
    request: ImageGenerationRequest
): Promise<ImageGenerationResponse> {
    return apiClient.post<ImageGenerationResponse>('/ai/generate/image', {
        prompt: request.prompt,
        provider: request.provider,
        model: request.model,
        style: request.style,
        width: request.width || 1024,
        height: request.height || 1024,
    }, {
        timeout: 120000,
    });
}

export async function generateVideo(
    request: VideoGenerationRequest
): Promise<VideoGenerationResponse> {
    return apiClient.post<VideoGenerationResponse>('/ai/generate/video', {
        prompt: request.prompt,
        provider: request.provider,
        model: request.model,
        width: request.width || 848,
        height: request.height || 480,
        duration: request.duration || 5,
        reference_image: request.reference_image,
    }, {
        timeout: 240000,
    });
}

export async function improveText(
    request: ImproveRequest
): Promise<ImproveResponse> {
    return apiClient.post<ImproveResponse>('/ai/improve', {
        content: request.content,
        platform: request.platform,
        provider: request.provider,
        model: request.model,
        instructions: request.instructions,
    });
}

export async function generateVariations(
    request: VariationsRequest
): Promise<VariationsResponse> {
    return apiClient.post<VariationsResponse>('/ai/generate/variations', {
        content: request.content,
        platform: request.platform,
        provider: request.provider,
        model: request.model,
        variations_count: request.variations_count || 3,
    });
}

export async function chat(request: ChatRequest): Promise<ChatResponse> {
    return apiClient.post<ChatResponse>('/ai/chat', {
        messages: request.messages,
        provider: request.provider,
        model: request.model,
        category: request.category,
        platform: request.platform,
    });
}

// ============================================================
// EXPORT
// ============================================================

export const aiApi = {
    getProviders,
    generateText,
    generateImage,
    generateVideo,
    improveText,
    generateVariations,
    chat,
};

export default aiApi;

File: postlio_backend/app/schemas/ai.py
# postlio_backend/app/schemas/ai.py

from pydantic import BaseModel, Field
from typing import Optional, List
from enum import Enum


# === ENUMS ===

class TextProviderEnum(str, Enum):
    GEMINI = "gemini"
    GROQ = "groq"


class ImageProviderEnum(str, Enum):
    POLLINATIONS = "pollinations"
    GEMINI = "gemini"  # NOWY - Gemini Image (Nano Banana)
    HUGGINGFACE = "huggingface"
    CLIPDROP = "clipdrop"  # Płatny


class CategoryEnum(str, Enum):
    FITNESS = "fitness"
    HEALTH = "health"
    BEAUTY = "beauty"
    COOKING = "cooking"
    BUSINESS = "business"
    TECHNOLOGY = "technology"
    TRAVEL = "travel"
    LIFESTYLE = "lifestyle"
    EDUCATION = "education"
    ENTERTAINMENT = "entertainment"


class ToneEnum(str, Enum):
    PROFESSIONAL = "professional"
    CASUAL = "casual"
    HUMOROUS = "humorous"
    INSPIRATIONAL = "inspirational"
    EDUCATIONAL = "educational"
    FRIENDLY = "friendly"


class PlatformEnum(str, Enum):
    FACEBOOK = "facebook"
    INSTAGRAM = "instagram"
    LINKEDIN = "linkedin"


class ImageStyleEnum(str, Enum):
    REALISTIC = "realistic"
    ARTISTIC = "artistic"
    CARTOON = "cartoon"
    MINIMALIST = "minimalist"
    VIBRANT = "vibrant"
    PROFESSIONAL = "professional"


# === TEXT REQUEST SCHEMAS ===

class GenerateTextRequest(BaseModel):
    topic: str = Field(..., min_length=3, max_length=500)
    platform: PlatformEnum
    provider: Optional[TextProviderEnum] = None
    model: Optional[str] = None
    category: Optional[CategoryEnum] = None
    tone: ToneEnum = Field(default=ToneEnum.PROFESSIONAL)
    language: str = Field(default="pl")
    include_hashtags: bool = Field(default=True)
    include_emoji: bool = Field(default=True)
    max_length: Optional[int] = None


class GenerateVariationsRequest(BaseModel):
    content: str = Field(..., min_length=10)
    platform: PlatformEnum
    provider: Optional[TextProviderEnum] = None
    model: Optional[str] = None
    variations_count: int = Field(default=3, ge=1, le=5)


class ImproveTextRequest(BaseModel):
    content: str = Field(..., min_length=10)
    platform: PlatformEnum
    provider: Optional[TextProviderEnum] = None
    model: Optional[str] = None
    instructions: Optional[str] = None


class ChatMessage(BaseModel):
    role: str = Field(..., pattern="^(user|assistant)$")
    content: str


class AIChatRequest(BaseModel):
    messages: List[ChatMessage]
    provider: Optional[TextProviderEnum] = None
    model: Optional[str] = None
    category: Optional[CategoryEnum] = None
    platform: Optional[PlatformEnum] = None


# === IMAGE REQUEST SCHEMAS ===

class GenerateImageRequest(BaseModel):
    prompt: str = Field(..., min_length=3, max_length=1000)
    provider: Optional[ImageProviderEnum] = None
    model: Optional[str] = None
    style: Optional[ImageStyleEnum] = None
    width: int = Field(default=1024, ge=256, le=2048)
    height: int = Field(default=1024, ge=256, le=2048)


# === RESPONSE SCHEMAS ===

class GeneratedTextContent(BaseModel):
    content: str
    platform: str
    provider: str
    model: Optional[str] = None
    hashtags: Optional[List[str]] = None


class GenerateTextResponse(BaseModel):
    success: bool
    data: GeneratedTextContent
    tokens_used: Optional[int] = None


class GenerateVariationsResponse(BaseModel):
    success: bool
    variations: List[str]
    provider: str


class AIChatResponse(BaseModel):
    success: bool
    message: str
    provider: str


class GeneratedImageContent(BaseModel):
    image_url: Optional[str] = None
    image_data: Optional[str] = None  # base64
    prompt: str
    prompt_translated: Optional[str] = None  # NOWE - dla auto-tłumaczenia
    prompt_enhanced: Optional[str] = None  # NOWE
    provider: str
    model: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None


class GenerateImageResponse(BaseModel):
    success: bool
    data: Optional[GeneratedImageContent] = None
    error: Optional[str] = None


class ProviderInfo(BaseModel):
    name: str
    display_name: Optional[str] = None  # NOWE
    available: bool
    is_free: Optional[bool] = True  # NOWE
    models: List[str]
    is_default: bool
    description: Optional[str] = None  # NOWE


class ProvidersListResponse(BaseModel):
    text_providers: List[ProviderInfo]
    image_providers: List[ProviderInfo]

